### PR TITLE
Fixes #10234 - Postgres SSL Certificate fix

### LINKED ISSUE #10234: Need to have Postgresql SSL Certificate option when sslmode is enabled
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

There is no option to add the below Postgresql SSL Certificate option when ssl is enabled and sslmode is verify-ca/require/verify-full 

	c.Options["sslrootcert"] 

	c.Options["sslkey"] 

	c.Options["sslcert"]

### Version

v3.3.8

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
enable SSL and SSL mode on postgresql db and try to integrate with argo workflow will get the below error:

time="2022-12-14T20:28:04.222Z" level=fatal msg="pq: connection requires a valid client certificate"
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}

time="2022-12-14T20:28:04.198Z" level=info msg="SSO enabled"
time="2022-12-14T20:28:04.202Z" level=info msg="Starting Argo Server" instanceID= version=latest+unknown
time="2022-12-14T20:28:04.202Z" level=info msg="Creating DB session"
time="2022-12-14T20:28:04.222Z" level=fatal msg="pq: connection requires a valid client certificate"
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

### PR DESCRIPTION
**Pull request by reddymh**
Signed-off-by: Rajshekar Reddy <reddymh@gmail.com>

Fixes #10234 

Please do not open a pull request until you have checked ALL of these:

* [x] Create the PR as draft .
* [x] Run `make pre-commit -B` to fix codegen and lint problems.
* [x] Sign-off your commits (otherwise the DCO check will fail).
* [x] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [x] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [x] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [x] Github checks are green.
* [] Once required tests have passed, mark your PR "Ready for review".

If changes were requested, and you've made them, dismiss the review to get it reviewed again.

### CODE REVIEW

**Comment by rohankmr414** [comment]
@reddymh have you tested this on your end?

File: Dockerfile [hunk 1]
```diff
@@ -101,6 +101,8 @@ FROM gcr.io/distroless/static as workflow-controller
 
 USER 8737
 
+WORKDIR /home/argo
+
```
Comment (rohankmr414) [inline Dockerfile:105]: do we really need this? can't we use absolute paths for files in `persist/sqldb/sqldb.go`

File: config/config.go [hunk 2]
```diff
@@ -231,8 +231,11 @@ func (c DatabaseConfig) GetHostname() string {
 
 type PostgreSQLConfig struct {
 	DatabaseConfig
-	SSL     bool   `json:"ssl,omitempty"`
-	SSLMode string `json:"sslMode,omitempty"`
+	SSL              bool                    `json:"ssl,omitempty"`
+	SSLMode          string                  `json:"sslMode,omitempty"`
+        CaCertSecret     apiv1.SecretKeySelector `json:"caCertSecret,omitempty"`
+        ServerCertSecret apiv1.SecretKeySelector `json:"serverCertSecret,omitempty"`
+        ServerKeySecret  apiv1.SecretKeySelector `json:"serverKeySecret,omitempty"`
```
Comment (rohankmr414) [inline config/config.go:238]: can you fix the lint here please https://github.com/argoproj/argo-workflows/actions/runs/3832454101/jobs/6522907398

File: persist/sqldb/sqldb.go [hunk 3]
```diff
@@ -53,9 +54,44 @@ func CreatePostGresDBSession(kubectlConfig kubernetes.Interface, namespace strin
 	}
 
 	if cfg.SSL {
-		if cfg.SSLMode != "" {
+		if cfg.SSLMode != "" && cfg.SSLMode != "disable" {
+			pgCertPath := "./pgcerts"
```
Comment (rohankmr414) [inline persist/sqldb/sqldb.go:58]: can we use the absolute path here and get rid of `WORKDIR /home/argo` from Dockerfile?

**Review by rohankmr414 — CHANGES_REQUESTED** [review]

**Comment by rohankmr414** [comment]
can you also sign off your commits for DCO https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal

**Comment by reddymh** [comment]
> 
yes . we have tested and running in prod as well

File: persist/sqldb/sqldb.go (hunk 3 above)
Comment (reddymh) [reply persist/sqldb/sqldb.go:58]: @rohankmr414  we ca use absolute path. I thought argo server already using workdir in Dockerfile so I thought same thing we can use for argo controller.

File: config/config.go (hunk 2 above)
Comment (reddymh) [reply config/config.go:238]: @rohankmr414  tried in visual studio and not getting any lint error but again I will cross check on lint errors.

Comment (rohankmr414) [reply config/config.go:239]: Can you run `make lint` in the repo directory? That'll fix it

File: persist/sqldb/sqldb.go (hunk 3 above)
Comment (rohankmr414) [reply persist/sqldb/sqldb.go:58]: I don't see any issues with it as well, it's just an extra change that can be avoided.

cc: @sarabala1979

File: Dockerfile (hunk 1 above)
Comment (sarabala1979) [reply Dockerfile:105]: WorkDir is needed to refer to where all the binary and files need to be copied. Can we config the path?

File: persist/sqldb/sqldb.go (hunk 3 above)
Comment (sarabala1979) [reply persist/sqldb/sqldb.go:58]: Can we add another element in structs to configure the path and have a default value too?

Comment (reddymh) [reply persist/sqldb/sqldb.go:58]: @sarabala1979  @rohankmr414  then we need to use constructor for this field in the struct

Comment (rohankmr414) [reply persist/sqldb/sqldb.go:58]: @reddymh yes you can add a field to the spec

**Review by rohankmr414 — APPROVED** [review]

File: persist/sqldb/sqldb.go (hunk 3 above)
Comment (reddymh) [reply persist/sqldb/sqldb.go]: @rohankmr414 @sarabala1979  added field to the spec and default vault if the variable is not configured in configmap or not defined.
Please review and approve the same

**Comment by reddymh** [comment]
Signed-off-by: Rajshekar Reddy <reddymh@gmail.com>

**Comment by reddymh** [comment]
> can you also sign off your commits for DCO https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal

I have added signoff but still showing DCO - action required

File: config/config.go [hunk 4]
```diff
@@ -232,8 +232,19 @@ func (c DatabaseConfig) GetHostname() string {
 
 type PostgreSQLConfig struct {
 	DatabaseConfig
-	SSL     bool   `json:"ssl,omitempty"`
-	SSLMode string `json:"sslMode,omitempty"`
+	SSL              bool                    `json:"ssl,omitempty"`
+	SSLMode          string                  `json:"sslMode,omitempty"`
+	CaCertSecret     apiv1.SecretKeySelector `json:"caCertSecret,omitempty"`
+	ServerCertSecret apiv1.SecretKeySelector `json:"serverCertSecret,omitempty"`
```
Comment (vladlosev) [inline config/config.go:238]: This field and the next are mapped to the `sslcert` and `sslkey` options in the PostgreSQL connection options. The PostgreSQL documentation [defines ](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS)them as client certificate and key filename, so naming them server cert and server key is incorrect.

**Review by sarabala1979 — APPROVED** [review]

File: config/config.go (hunk 4 above)
Comment (reddymh) [reply config/config.go]: @vladlosev changed the  ServerCertSecret and ServerKeySecret as ClientCertSecret and ClientCertSecret and will change respective helm chart as well.

Comment (reddymh) [reply config/config.go]: cc: @sarabala1979  @rohankmr414

**Review by vladlosev — APPROVED** [review]
FWIW

**Comment by rohankmr414** [comment]
> I have added signoff but still showing DCO - action required
@reddymh you'll have to sign off all your commits, you can also squash all your commits into one and then sign off the commit, you can also go to the `details` of the failing check and click the button to accept the DCO

**Comment by reddymh** [comment]
@sarabala1979 @rohankmr414 @vladlosev  Can we merge the PR ?

**Comment by reddymh** [comment]
> > I have added signoff but still showing DCO - action required
> > @reddymh you'll have to sign off all your commits, you can also squash all your commits into one and then sign off the commit, you can also go to the `details` of the failing check and click the button to accept the DCO

@rohankmr414 Can we merge the PR?

**Comment by tooptoop4** [comment]
@reddymh do u think this caused https://github.com/argoproj/argo-workflows/issues/10498?

**Pull request merged**